prow/tide/search.go

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tide
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	githubql "github.com/shurcooL/githubv4"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	timeFormatISO8601 = "2006-01-02T15:04:05Z"
+)
+
+type searchExecutor func(start, end time.Time) ([]PullRequest, int /*true match count*/, error)
+
+func newSearchExecutor(ctx context.Context, ghc githubClient, log *logrus.Entry, q string) searchExecutor {
+	return func(start, end time.Time) ([]PullRequest, int, error) {
+		q = fmt.Sprintf("%s %s", q, dateToken(start, end))
+		vars := map[string]interface{}{
+			"query":        githubql.String(q),
+			"searchCursor": (*githubql.String)(nil),
+		}
+		var totalCost, remaining int
+		var totalMatches int
+		var ret []PullRequest
+		for {
+			sq := searchQuery{}
+			if err := ghc.Query(ctx, &sq, vars); err != nil {
+				return nil, 0, err
+			}
+			totalCost += int(sq.RateLimit.Cost)
+			remaining = int(sq.RateLimit.Remaining)
+			totalMatches = int(sq.Search.IssueCount)
+			// If the search won't return all results, abort.
+			if totalMatches > 1000 {
+				return nil, totalMatches, nil
+			}
+			for _, n := range sq.Search.Nodes {
+				ret = append(ret, n.PullRequest)
+			}
+			if !sq.Search.PageInfo.HasNextPage {
+				break
+			}
+			vars["searchCursor"] = githubql.NewString(sq.Search.PageInfo.EndCursor)
+		}
+		log.Debugf("Search for query \"%s\" returned %d PRs and cost %d point(s). %d remaining.", q, len(ret), totalCost, remaining)
+		return ret, totalMatches, nil
+	}
+}
+
+func (q searchExecutor) search() ([]PullRequest, error) {
+	prs, _, err := q.searchRange(time.Time{}, time.Now())
+	return prs, err
+}
+
+func (q searchExecutor) searchSince(t time.Time) ([]PullRequest, error) {
+	prs, _, err := q.searchRange(t, time.Now())
+	return prs, err
+}
+
+func (q searchExecutor) searchRange(start, end time.Time) ([]PullRequest, int, error) {
+	prs, count, err := q(start, end)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	if count <= 1000 {
```

Comment:
Can we do anything to read how many GitHub says there are instead of ingesting 1000 PR results and throwing them away? Do we think that will have any impact on startup or response time, or performance? How much of a memory footprint does this introduce?

Reply:
We never ingest 1000 results and throw them away. With the first page of 100 results we also get the true count of the number of PRs that match the query. If this number is more than 1000 we know that we won't be able to use the partial results so we abort and just return the true count: https://github.com/kubernetes/test-infra/pull/9263/files#diff-182165d8d4fb3d1bc442acfed435c7d2R53